codec/src/main/java/io/netty/handler/codec/protobuf/nano/ProtobufEncoderNano.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.protobuf.nano;
+
+import com.google.protobuf.nano.CodedOutputByteBufferNano;
+import com.google.protobuf.nano.MessageNano;
+
+import java.util.List;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
+import io.netty.handler.codec.LengthFieldPrepender;
+import io.netty.handler.codec.MessageToMessageEncoder;
+
+import static io.netty.buffer.Unpooled.wrappedBuffer;
+
+/**
+ * Encodes the requested <a href="http://code.google.com/p/protobuf/">Google
+ * Protocol Buffers</a> {@link MessageNano} into a
+ * {@link ByteBuf}. A typical setup for TCP/IP would be:
+ * <pre>
+ * {@link ChannelPipeline} pipeline = ...;
+ *
+ * // Decoders
+ * pipeline.addLast("frameDecoder",
+ *                  new {@link LengthFieldBasedFrameDecoder}(1048576, 0, 4, 0, 4));
+ * pipeline.addLast("protobufDecoder",
+ *                  new {@link ProtobufDecoderNano}(MyMessage.getDefaultInstance()));
+ *
+ * // Encoder
+ * pipeline.addLast("frameEncoder", new {@link LengthFieldPrepender}(4));
+ * pipeline.addLast("protobufEncoder", new {@link ProtobufEncoderNano}());
+ * </pre>
+ * and then you can use a {@code MyMessage} instead of a {@link ByteBuf}
+ * as a message:
+ * <pre>
+ * void channelRead({@link ChannelHandlerContext} ctx, Object msg) {
+ *     MyMessage req = (MyMessage) msg;
+ *     MyMessage res = MyMessage.newBuilder().setText(
+ *                               "Did you say '" + req.getText() + "'?").build();
+ *     ch.write(res);
+ * }
+ * </pre>
+ */
+@ChannelHandler.Sharable
+public class ProtobufEncoderNano extends MessageToMessageEncoder<MessageNano> {
+    @Override
+    protected void encode(
+            ChannelHandlerContext ctx, MessageNano msg, List<Object> out) throws Exception {
+        byte[] buffer = new byte[msg.getSerializedSize()];
+        CodedOutputByteBufferNano cobbn = CodedOutputByteBufferNano.newInstance(buffer);
+        msg.writeTo(cobbn);
+        out.add(wrappedBuffer(buffer));
```

Comment:
i'd like that but i see no way of advancing the writerIndex of the buffer.
So as i see it there are 3 possibilites:
* use current implementation with unpooled heap array
* create pooled ByteBuf, use its heap array like you proposed and then wrap it in an unpooled ByteBuf. I don't know what implications this will have. We need to release the Pooled ByteBuf right? if we do that it will create race between writing the buffer to socket and reuse of the heap memory by pool?
* as above but writing the array into the ByteBuf. Could this be optimized away? does the underlying copy implementation compare src and dst addresses and realise they're the same?